### DISCUSSION TITLE
Tile pore network array into itself and merge with another network at a different scale

### DISCUSSION THREAD

**Discussion by baq99** [PMEAL/OpenPNM#2167] [Q&A] [answered]
Hello everyone.  I have hit another snag using OpenPNM and I'm not sure how to resolve it.  I am currently working on two OpenPNM projects simultaneously, each containing a different network with associated geometry at different scales.  I have project 1 (at a scale of 1) containing pn1 and geo1, and project 2 (at a scale of 0.1) containing pn2 and geo2.  

What I would like to do is something similar to the "tile" command for numpy arrays.  I would like to expand project 2 ten times, repeating it's pore information.

Similar to:
`a = np.array([0, 1, 2])`
`np.tile(a, 10)`
`array([0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2])`

The intention being to extrapolate project 2 to the same size as project 1.  I am then planning to merge the two networks together into one.  What I would like to show is a prediction of the pore volume contribution from the smaller network at the size of the larger network and the predicted total pore volume and pore network distribution of the combined and extrapolated networks.

Unfortunately I can't figure out how to tile the project 2 array and merge project 1 and 2 thereafter.

Would this be possible?

**Comment by jgostick** [comment, marked as answer]
Hi @baq99 
For part 1 of your mission we have a function called 'openpn.topotools.stitch' that may do the trick for you.  It was really designed to stitch *two* networks together so I'm not sure how it'll fair with N by N networks.  

For part 2, I cannot quite picture what you mean by merging the two networks into one.  Would these be superimposed somehow?  

Also, on a related note, since your project seems to invovle a lot of volume calculations, be careful about the overlap between pores and throats.  Depending on how to you define the geometry there can either be a lens created by the overlap, or gap between the pore and end of the throat.  In the former case you may count the volume twice, while in the later case you will be missing some volume.  There are models in openpnm.models.geometry.throat_volume called pendular_ring and lens, which help address this.

**Reply by baq99** [reply]
Hi Jgostick,
Thank you for your answer.  

For part 1, can I confirm you're suggesting I duplicate the network and then stitch them together ten times as a solution for increasing the size of the network?  I didn't think of this as an option but I'll give it a try.  Thank you. 

For part 2, yes I was hoping that I could somehow superimpose the two networks together giving me a much greater range of scale.

**Reply by jgostick** [reply]
For part 1 you can do it by creating 2 'small' networks, PN1 and PN2.  You the 'stitch' PN2 to PN1 recursively.  Openpnm will do the duplication internally.  The function will return PN1 to you with the new pores added, so you just keep passing it back in.  You'll just have to manually translate the pores of PN2 each time to be in the correct position.  We use a 'nearest neighbor/distance map' look-up to determine which pores on PN1 get connect to which pores on PN2.  

For part 2, I'm afraid I don't see a clear path forward.  We have done some 'multiscale' networks, but starting with the 'small' network and replacing groups of pores by larger ones:  https://doi.org/10.1016/j.cej.2017.07.139.  This worked pretty well.  We had considered the opposite approach, where you start with a coarse network and 'subdivide' large pores into smaller ones, but it wasn't as satisfactory.  The function can be found in ``openpnm.topotools.subdivide``, but I am not confident that it's robust.

**Reply by baq99** [reply]
That's great information thank you very much for taking the time to explain.  

I think for the time being then, I will have to produce values from the small and large networks separately and stitch the results after the fact, rather than combining the two networks before generating the output.

However I will have a try at scaling up the small porosity to match that of the large as this will likely make my job easier when it comes to the results.

Thank you for your detailed reply.  It's been very helpful.
